protocol/streaming/grpc/grpc_streaming_manager.go

```diff
@@ -9,30 +13,125 @@ var _ types.GrpcStreamingManager = (*GrpcStreamingManagerImpl)(nil)
+	nextId                 uint32
+}
+
+// OrderbookSubscription represents a active subscription to the orderbook updates stream.
+type OrderbookSubscription struct {
+	clobPairIds []uint32
+	srv         clobtypes.Query_StreamOrderbookUpdatesServer
+	finished    chan bool
 }
 
 func NewGrpcStreamingManager() *GrpcStreamingManagerImpl {
-	return &GrpcStreamingManagerImpl{}
+	return &GrpcStreamingManagerImpl{
+		orderbookSubscriptions: make(map[uint32]*OrderbookSubscription),
+	}
 }
 
 func (sm *GrpcStreamingManagerImpl) Enabled() bool {
 	return true
 }
 
 // Subscribe subscribes to the orderbook updates stream.
+// This function returns a channel that is used to signal termination when an error occurs.
 func (sm *GrpcStreamingManagerImpl) Subscribe(
 	req clobtypes.StreamOrderbookUpdatesRequest,
 	srv clobtypes.Query_StreamOrderbookUpdatesServer,
 ) (
 	finished chan bool,
 	err error,
 ) {
-	return nil, nil
+	clobPairIds := req.GetClobPairId()
+
+	// Perform some basic validation on the request.
+	if len(clobPairIds) == 0 {
+		return nil, clobtypes.ErrInvalidGrpcStreamingRequest
+	}
+
+	finished = make(chan bool)
+	subscription := &OrderbookSubscription{
+		clobPairIds: clobPairIds,
+		srv:         srv,
+		finished:    finished,
+	}
+
+	sm.Lock()
+	defer sm.Unlock()
+
+	sm.orderbookSubscriptions[sm.nextId] = subscription
+	sm.nextId++
+
+	return finished, nil
 }
 
 // SendOrderbookUpdates groups updates by their clob pair ids and
 // sends messages to the subscribers.
 func (sm *GrpcStreamingManagerImpl) SendOrderbookUpdates(
-	updates *clobtypes.OffchainUpdates,
+	offchainUpdates *clobtypes.OffchainUpdates,
 ) {
+	// Group updates by clob pair id.
+	updates := make(map[uint32]*clobtypes.OffchainUpdates)
+	for _, message := range offchainUpdates.Messages {
+		clobPairId := message.OrderId.ClobPairId
+		if _, ok := updates[clobPairId]; !ok {
+			updates[clobPairId] = clobtypes.NewOffchainUpdates()
+		}
+		updates[clobPairId].Messages = append(updates[clobPairId].Messages, message)
+	}
+
+	// Unmarshal messages to v1 updates.
+	v1updates := make(map[uint32][]ocutypes.OffChainUpdateV1)
+	for clobPairId, update := range updates {
+		v1update, err := GetOffchainUpdatesV1(update)
+		if err != nil {
+			panic(err)
+		}
+		v1updates[clobPairId] = v1update
+	}
+
+	sm.Lock()
+	defer sm.Unlock()
+
+	// Send updates to subscribers.
+	idsToRemove := make([]uint32, 0)
+	for id, subscription := range sm.orderbookSubscriptions {
+		for _, clobPairId := range subscription.clobPairIds {
+			if updates, ok := v1updates[clobPairId]; ok {
+				if err := subscription.srv.Send(
+					&clobtypes.StreamOrderbookUpdatesResponse{
+						Updates:  updates,
+						Snapshot: false,
+					},
+				); err != nil {
+					idsToRemove = append(idsToRemove, id)
+					break
+				}
+			}
+		}
+	}
+
+	// Clean up subscriptions that have been closed.
+	for _, id := range idsToRemove {
+		sm.orderbookSubscriptions[id].finished <- true
+		delete(sm.orderbookSubscriptions, id)
```

Comment:
Are we sure that this delete will properly clean up the nested inner data structures? `grpc.ServerStream` might have a close function?

Reply:
Seems like there's no close function 🆗

Reply:
yeah wasn't able to find a close function either 

not really sure if we need to do `ctx, cancelCtx := context.WithCancel(ctx)` to cancel the context, since the connection is likely already broken and the context is already done 🤔

Reply:
https://github.com/grpc/grpc-go/blob/v1.60.x/stream.go#L262-L278
```
	// Possible context leak:
	// The cancel function for the child context we create will only be called
	// when RecvMsg returns a non-nil error, if the ClientConn is closed, or if
	// an error is generated by SendMsg.
```
I'm assuming an error will be generated by SendMsg in this case so the cancel function will be called.